add code_prettify extension

jfbercher opened the pull request:
- remove `yapf_ext` and add `code_prettify`
- `code_prettify`  enhances yapf_ext and adds support for R and javascript in addition to Python


# A Code Prettifier

This extension reformats/prettifies code in a notebook's code cell. 
It currently supports R, javascript and Python code. 
Under the hood, it uses the [YAPF](https://github.com/google/yapf) Python module to reformat Python code (python2/python3 kernels), [formatR](http://yihui.name/formatR/) for R code (ir kernel) and [js-beautify](https://github.com/beautify-web/js-beautify) for javascript ([ijavascript](http://n-riesco.github.io/ijavascript/) kernel). 

Other languages may be added in the future. Actually, given that there are more than 50 kernels available for Jupyter it does not easy possible to support all of them. The current implementation uses a call to the current kernel to reformat the code. Thus the actual prettyfier package has to be written in the current kernel language. 

**pre-requisites:** of course, you must have some of the corresponding packages installed:

- for Python 
```
pip install yapf [--user]
``` 
- for R
```
install.packages("formatR", repos = "http://cran.rstudio.com")
``` 
- for ijavascript (*in the root of your user tree = ~*)
```
npm install js-beautify
``` 

Then the extension provides

- a toolbar button
- a keyboard shortcut for reformatting the current code-cell (default: Ctrl-L)

Syntax shall be correct. The extension will also point basic syntax errors. 
![demo-py](https://cloud.githubusercontent.com/assets/7596356/17790418/a7af66e0-6596-11e6-88a9-b0f136978536.gif)
![demo-r](https://cloud.githubusercontent.com/assets/7596356/17790423/ac7782ac-6596-11e6-8901-66c69c07bf9b.gif)
![demo-jv](https://cloud.githubusercontent.com/assets/7596356/17790427/b051207c-6596-11e6-9482-39a3d8cf06f6.gif)


Installation
------------

If you use [jupyter-contrib-nbextensions](https://github.com/ipython-contrib/jupyter_contrib_nbextensions), proceed as usual. 

Otherwise, you can still install/try the extension from my personal repo, using
```
jupyter nbextension install https://github.com/jfbercher/code_prettify/archive/master.zip --user
jupyter nbextension enable code_prettify-master/code_prettify
```


History: 
---------

- @jfbercher, august 14, 2016, first version. 
- @jfbercher, august 19, 2016, second version, introducing support for R language and javascript. Change extension name from `yapf_ext` to `code_prettify`

janschulz commented on src/jupyter_contrib_nbextensions/nbextensions/code_prettify/code_prettify.js:
```diff
@@ -0,0 +1,199 @@
+            exec: yapf_format,
+            post_exec: ''
+        },
+        ir: {
```
I haven't tested this, but I have no kernel named ir, but multiple IRkernel environments named like `ir32` and  `ir33` and also conda environments which are named like `conda_r_envname`.

So IMO this should use the language tag and not the kernel name itself.

jfbercher replied:
ok  indeed, `ir` is the default name but users can specify a [different name at install](https://github.com/IRkernel/IRkernel). There is no language tag accessible in the kernel object, so no indication of what is the underlying language. 
I have added a map connecting names with the associated language and use the language as the key for selecting the prettifier.  For  other names defined by users, I am afraid that they have to be added by hand (or perhaps via some parameters in the yaml?)

janschulz commented:
Sorry, `conda_r_envname`  is just an example: I use https://github.com/Cadair/jupyter_environment_kernels/ to get all conda environments with a kernel inside to be included in the list of available kernels, so I have the following kernels: `conda_dev_35, conda_env_manager_35, conda_monte-carlo, conda_matplotlib_build, conda_env_manager_27, conda_ipcluster, conda_r_r33, conda_py35, conda_sc, conda_jupyter-notebook`.  I meant that the default kernelspec list (http://localhost:8888/api/kernelspecs) contains a `language` tag: 

```json
{  
   "kernelspecs":{  
      "conda_dev_35":{  
         "name":"conda_dev_35",
         "spec":{  
            "env":{  

            },
            "argv":[  
               "C:\\portabel\\miniconda\\envs\\dev_35\\python.exe",
               "-m",
               "IPython.kernel",
               "-f",
               "{connection_file}"
            ],
            "language":"python",
            "display_name":"~env (conda_dev_35)"
         },
         "resources":{  
            "logo-64x64":"/kernelspecs/conda_dev_35/logo-64x64.png",
            "logo-32x32":"/kernelspecs/conda_dev_35/logo-32x32.png"
         }
      },
      "conda_env_manager_35":{  
         "name":"conda_env_manager_35",
         "spec":{  
            "env":{  

            },
            "argv":[  
               "C:\\portabel\\miniconda\\envs\\env_manager_35\\python.exe",
               "-m",
               "IPython.kernel",
               "-f",
               "{connection_file}"
            ],
            "language":"python",
            "display_name":"~env (conda_env_manager_35)"
         },
         "resources":{  
            "logo-64x64":"/kernelspecs/conda_env_manager_35/logo-64x64.png",
            "logo-32x32":"/kernelspecs/conda_env_manager_35/logo-32x32.png"
         }
      },
[...]
```

jfbercher commented:
@janschulz 
ok I wasn't aware of these conda_envs. I have updated the source to read the lists of kernels from the server api and extracted the language key. Now I must go back to some other duties.

janschulz commented on src/jupyter_contrib_nbextensions/nbextensions/code_prettify/code_prettify.js:
```diff
@@ -31,7 +31,7 @@ define(function(require, exports, module) {
             post_exec: ''
         },
-        r: {   // intentionnaly in lower case
+        r: { // intentionnaly in lower case
```
typo: intentionally

janschulz commented on src/jupyter_contrib_nbextensions/nbextensions/code_prettify/code_prettify.js:
```diff
@@ -157,52 +157,42 @@ define(function(require, exports, module) {
+        kName = Jupyter.notebook.kernel.name;
+        console.log("kName",kName)
+        // get the list of user kernels and extract the language tag
+        $.getJSON(Jupyter.notebook.kernel.ws_url.replace("ws:", "http:") + '/api/kernelspecs',
```
There is also `/api/kernelspecs/<name>` https://github.com/jupyter/notebook/blob/master/notebook/services/kernelspecs/handlers.py#L86

What I don't understand is why this information isn't available already? the editor component knows what kind of highlighting it should do, so that information should be available somewhere in the UI...

jfbercher replied:
Yes indeed! Trivial: IPython.notebook.metadata.kernelspec.language

jcb91 commented:
are we ok to merge this?

janschulz commented on src/jupyter_contrib_nbextensions/nbextensions/code_prettify/README.md:
```diff
@@ -0,0 +1,53 @@
+Other languages may be added in the future. Actually, given that there are more than 50 [kernels](https://github.com/ipython/ipython/wiki/IPython-kernels-for-other-languages) available for Jupyter it does not easy possible to support all of them. The current implementation uses a call to the current kernel to reformat the code. Thus the actual prettyfier package has to be written in the current k ...
+
+**pre-requisites:** of course, you must have some of the corresponding packages installed:
+
```
`.. in the kernel you work in.`

jcb91 commented:
So this is already becoming a bit of a tangle of escaping multiple things in multiple places to convert between javascript strings and kernel-language strings. I suggest:

* JSON strings are simple, and every kernel has to be capable of encoding & decoding JSON in order to pass messages. So, it makes sense to use them as the basis for interchange. See the following image (from [this article](http://timelessrepo.com/json-isnt-a-javascript-subset)) for a neat summary of what constitutes a valid JSON string: ![](http://stuff.judofyr.net/json/json-string.gif). This makes for relatively simple conversion to other languages' string literal representations.
* For conversion from frontend to kernel, call `JSON.stringify(cell_text).replace(/\u2028/g, '\\u2028').replace(/\u2029/g, '\\u2029');` to get a javascript-valid JSON string (see [this MDN article](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#Issue_with_plain_JSON.stringify_for_use_as_JavaScript) for why the replacements are a good idea). Then potentially kernel-specific replacements can be performed to convert to a valid string literal for the given kernel. I suspect, however, that many kernels won't need much extra work. In python, for example, JSON unicode escape sequences are also valid python unicode escapes.
* For conversion from kernel to frontend, perhaps it would be better to get the kernels to print a JSON representation of the output which avoids the kernel constructing its own string literal (e.g. adding quotes, prefixes etc), and then the frontend can simply collect the appropriate iopub message(s) and use them directly, rather than having to worry about unescaping the kernel's idea of a string literal.

jfbercher commented:
Thanks for the input and useful pointers. Indeed I finished in using JSON.stringify (see around lines 133-135) in last commits. I have no more time to implement the remaining for now. So this will wait some time.

jcb91 commented:
@jfbercher Ah yes, so you did! Ok, I think then that we merge now, and later updates can be applied as subsequent separate PRs as appropriate.
